marlowe-runtime/docs/filtered-chain-sync-spec-1.0.md

```diff
@@ -0,0 +1,375 @@
+requirements for its protocols is said to be "conditionally compliant."
+
+### Terminology
+
+The following terminology is used to refer to the roles of participands in, and
+objects of, the protocol.
+
+- connection: A transport layer channel over which a client and a server exchange messages.
+- client: The role played by the participant which sends queries and receives chain data.
+- server: The role played by the participant which responds to queries maintains chain data.
+- state: A member of the set of states specified by the state machine of the protocol.
+- message: A member of the set of state transitions specified by the state maching of the protocol.
+- payload: Data associated and sent with a message.
+- query: Part of the payload of a query next message which the client sends to the server to tell it what to send next.
+- error: A protocol level error which is triggered when one party sends an invalid message. Results in the termination of the connection.
+- rejection: An application level error which is triggered when the server rejects a syntactically valid, semantically invalid query sent by the client. Does not result in the termination of the connection.
+- roll forward: A message sent by the server to the client in response to a query.
+- result: Part of the payload of a roll forward message which contains the result of a query.
+- roll backward: A messagge sent by the server to the client when the client's position has been rolled back.
+- tip: The latest chronological point in the blockchain
+- point: A point in the blockchain up to and including the tip
+- position: The point where a client is currently "focused."
+- schema version: A named and numberedversion of the query schema supported by the parties of the protocol. Used in the handshake to verify that the server supports the query schema supported by the client.
+- handshake: A proceedure used to establish a connection which ensures the client and server understand the same queeries.
+- agency: Used to describe which role is permited to send a message in a given state.
+
+### Operation
+
+The Filtered Chain Sync Protocol is a stateful client-server protocol. When a
+client establishes a connection with a server, a handshake is performed. If the
+handshake succeeds, the idle state is reached. From the idle state, the client
+has the agency to send a query or terminate the connection. If the client
+terminates the connection, by sending a `Done` message, then the server closes
+the connection and the protocol terminates. If the client sends a query, by
+sending a `QueryNext` message, the server can choose to respond immediately to
+the query, reject the query with a rejection reason, or ask the client to wait
+for a response. When the client receives a response or a rejection, it returns
+to the idle state, and the loop repeats.
+
+A response can be a roll forward or roll backward message. If the response is a
```

Comment:
Does the state machine need more failure transitions? For instance, what if the server accepts the query and transitions to the `MustWait` state, but there is a server-side failure (e.g., lost connect to the node or database)? It appears that the state machine has already committed to a roll forward or roll back that it might not be able to service. Should there be an option to report a failure, either retryable or not?

Reply:
In principle, the server could lose and regain its connection to the node while in a `Next any` state. If it can reconnect and resume from where it was, things proceed as if no interruption occurred. If it fails to find its old position, the client may receive a `RollBackward` message. Either way, this event is opaque to the client. I don't think the protocol should represent cases like these; we risk a leaky abstraction if it does.

In the case of a database connection error, the server could either try to reconnect or (more likely) die. If the server dies, it will server the connection at the transport layer. The client may try to re-establish the lost connection as it sees fit.

Reply:
Sounds good.